api/valor_api/backend/metrics/text_generation.py

```diff
@@ -380,6 +459,58 @@ def _compute_text_generation_metrics(
                     )
                     for metric in bleu_metrics
                 ]
+
+            if is_ContextPrecision_enabled:
+                assert client
+                for (prediction_context_list, groundtruth_list) in zip(
+                    list_of_prediction_context_lists, references
+                ):
+                    score = 0
+                    for groundtruth in groundtruth_list:
+                        score = max(
+                            score,
+                            client.context_precision(
+                                query=datum_text,
+                                context_list=prediction_context_list,
+                                groundtruth=groundtruth,
+                            ),
```

Comment:
I don't have a strong opinion. as a user, the least confusing approach (IMO) would be to return individual yes/no responses for each groundtruth/context. if there are multiple contexts and groundtruths, then I'd want to understand which specific context is useful for each specific groundtruth (i.e., no aggregation). if you don't aggregate, then the user can answer questions like "which pieces of context were useful for both groundtruths?" by themselves

totally up to you how to approach these scores. I'd just pick one approach and try to gather feedback to refine it

Reply:
This seems like a likely edge case.

I think there could be two interesting approaches to this:
1. First off, this seems like it could be similar to the object detection problem that's solved by average-precision. 
2. If precision is determined by contextual coverage of the groundtruths then defining the score to be `max(max(scores_gt1), max(scores_gt2))` could also be considered?

Reply:
I looked into the average precision approach per Charles' suggestion but that approach doesn't apply as the verdicts for each context, ground truth pair in context precision are just "yes" or "no". I also think there are pros to Nick's suggestion of returning the verdicts to the user instead of aggregating into a score for the user, however that would be a significant departure from how all of the text generation metrics work now. I wouldn't want to make that change unless we were changing it everywhere / adding verdict information to all of the metrics. We should wait on user feedback or at the very least a future PR for that.
I implemented ContextPrecision by getting an aggregate verdict for each context (aggregating over the ground truths), then computing the precision @ k from the aggregate verdicts. For the purposes of a single score, this approach makes the most sense. This metric is a precision, meaning we want to know if each retrieved context is "good" or "bad", and a retrieved context should be considered "good" if it supports any ground truth answer.

Reply:
> We should wait on user feedback or at the very least a future PR for that.

I'm on board with that idea